### DISCUSSION TITLE
Tabulature reading

### DISCUSSION THREAD

**Discussion by troger19** [Audiveris/audiveris#754] [Q&A] [answered]
Hello
I am wondering, why reading guitar tabulature partituras is not supported? On the first sight it seems programatically easier than reading classical notes. 
Thank you

**Comment by hbitteur** [comment]
Ah, ah!
What is funny is that Audiveris was initially developed to allow the recognition of classical notation. Why? Simply because I personally can read tablatures but not classical notation...

OK, perhaps it's time to give tablature recognition a second chance.
Especially now that OCR is rather well integrated in the application.
Could you point us to some collections of representative tablatures?

**Reply by troger19** [reply]
Yes, you are right, classical notation are harder to read and especially for guitar, when 1 note can be play on diffefrent frets. Also there are different types of tabulatures (for guitar) which has poor quality considering articulation and timing. That can make it very difficult to read (OCR) when only fret numbers are in single tact and duration is reflected with gaps or distance between the numbers. However there are also "high quality" tabs which includes tact signature, stems and all other stuff necesary to produce 1:1 altetrnative to regular notation score. 
Regarding your question, I think there are many options nowdays, for example download big database of guitar pro files, which provideds validatiton with no cost. 
If I can ask technical question, is Audiveris using some methods of current AI boom? Cause even if I put simple picture of 1 bar, no common AI GPTs recognizes tabulature correctly. In that case, good old OCR is the only choice? Seems quite obsolete in these days. 
I am doing some experiments using Tesseract, after removing horizontal lines, it can read fret numbers quite OK, but then all the logic needs to be developed around which fret was on which string, relative distance calculation, bar bondaries. I would rather see something like training, or classification and labeling proces, than OCR, but dont know even where to start :)

My goal is to be able to make a screenshot from youtube video, where player provides tabulature on the screen and have it transcribed to guitar pro for personal exercise. (dont advice me to do it manually,  to subscribe to channel, , or Herohero.. and yes, if this wouuld be possible in the future, then the people start to put some watermark over the tabulature, but thats a diffefrent story)
Sorry for long comment

**Comment by hbitteur** [comment, marked as answer]
I was used to Marcel Dadi's books of good quality with fret + stems + beams. 
But tablatures can be very different from one editor to the other.
This the reason why I'm asking you to provide pointers to the kinds of tablatures you are aiming at.
And then, we'll be able to evaluate the global feasability of the recognition task.

The current Audiveris version uses various methods to process a partition, depending on the items to handle. For instance, note heads use template matching, beams use image closing, texts use an external OCR software,  stand alone symbols use a neural network trained on representative glyphs, etc.

Regarding tablatures, I think OCR would not be the key component, because it is meant to detect lines of words, not isolated numbers.
Audiveris is rather good at detecting/erasing staff lines. Therefore, if the fret numbers don't touch the line pixels, they could easily be extracted and then recognized by the neural network. 
I need concrete examples to further investigate.

Regarding AI, Audiveris current neural network is a shallow one. Using deep learning techniques, I developed a patch classifier last year, able to detect and classify a potential music symbol around any given point of the input image. It worked but was too slow, the processing of a full page via a sliding window would have taken several days! :-)
I need to work on something like a single-shot detection (SSD), similar to face detection in a crowd, so that all musical symbols in the input image would be detected globally. This is the plan...

**Reply by troger19** [reply]
Thank you Hervé for your erudite explanation and broaden my horizons. Regarding tabs, at the beginning I am aiming f.i. for this kind of tab.
![image](https://github.com/user-attachments/assets/7a8566f5-07f4-48ee-8d07-872318693644)
